src/django_components/util/template_tag.py

```diff
@@ -87,9 +35,106 @@ class TagSpec(NamedTuple):
     - and treated as `only=False` and `required=False` if omitted
     """
 
+    def copy(self) -> "TagSpec":
+        sig_parameters_copy = [param.replace() for param in self.signature.parameters.values()]
+        signature = inspect.Signature(sig_parameters_copy)
+        flags = self.flags.copy() if self.flags else None
+        return self.__class__(
+            signature=signature,
+            tag=self.tag,
+            end_tag=self.end_tag,
+            flags=flags,
+        )
+
+    def validate_params(self, params: List["TagParam"]) -> Tuple[List[Any], Dict[str, Any]]:
+        """
+        Validates a list of TagParam objects against this tag spec's function signature.
+
+        The validation preserves the order of parameters as they appeared in the template.
+
+        Args:
+            params: List of TagParam objects representing the parameters as they appeared
+                   in the template tag.
+
+        Returns:
+            A tuple of (args, kwargs) containing the validated parameters.
+
+        Raises:
+            TypeError: If the parameters don't match the tag spec's rules.
+        """
+
+        # Create a function with this signature that captures the input and sorts
+        # it into args and kwargs
+        def validator(*args: Any, **kwargs: Any) -> Tuple[List[Any], Dict[str, Any]]:
+            # Let Python do the signature validation
+            bound = self.signature.bind(*args, **kwargs)
```

Comment:
Third part of the validation magic is that inside the `validator`, we use the given `TagSpec.signature` to figure out which parameters are args and kwargs.

Distinguishing between args and kwargs is important, because some of the Component features require this distinction.
- E.g. for the type hints when passing inputs to `Component.render()`, args and kwargs have to be defined separately due to Python typing limitations.
- Or the [feature about component defaults](https://github.com/EmilStenstrom/django-components/issues/881), that will likely (or at least at first) apply only to kwargs.